### DISCUSSION TITLE
Clean Output Directory

### DISCUSSION THREAD

**Discussion by squeaky-joe** [oven-sh/bun#4643] [Ideas]
Cleaning an output directory for compiled TypeScript files is undoubtedly a valuable feature for the Bun compiler, bringing a multitude of benefits to developers. This feature not only simplifies the development workflow but also promotes cleaner, more efficient codebases. Let's delve into why cleaning an output directory is such an awesome addition to the Bun compiler.

Maintaining Code Hygiene: Over time, as code evolves, numerous TypeScript files may be added, modified, or removed. Without a cleaning feature, the output directory can quickly become cluttered with outdated JavaScript files, sourcemaps, and other artifacts. This clutter can make it challenging to track down issues or maintain an organized codebase. A built-in cleaning mechanism ensures that only the relevant and up-to-date files remain, fostering code hygiene and reducing confusion.

Faster Development Iterations: During development, it's common to make frequent code changes and recompile TypeScript files. Without cleaning, the old compiled files may linger in the output directory, potentially leading to confusion or even runtime errors if outdated code is inadvertently executed. A clean output directory guarantees that each compilation starts from a clean slate, reducing the risk of unexpected behavior and speeding up development iterations.

Disk Space Optimization: As a project evolves, the output directory can accumulate a significant amount of redundant files. This not only clutters the workspace but also consumes valuable disk space. A cleaning feature helps optimize disk space usage by removing files that are no longer needed, ensuring that your development environment remains efficient and clutter-free.

Enhanced Collaboration: When working on a project with a team, it's crucial to have a consistent and clean codebase. A clean output directory simplifies collaboration by ensuring that everyone is working with the same set of compiled files. This consistency minimizes potential conflicts and discrepancies between team members, promoting smoother collaboration.

Automation and Convenience: Manually cleaning the output directory can be a tedious and error-prone task. Automating this process within the Bun compiler streamlines development and reduces the chance of human error. Developers can focus on writing code while the compiler takes care of keeping the output directory tidy.

Debugging and Troubleshooting: When troubleshooting issues in a TypeScript project, having a clean output directory is invaluable. It eliminates the possibility of conflicts arising from outdated or cached files, making it easier to pinpoint and fix problems. Clean output also ensures that sourcemaps and debugging information accurately reflect the current state of the codebase.

Customization and Control: A well-designed cleaning feature can provide developers with flexibility. They can configure which files to retain, which to delete, and specify any exceptions or additional cleanup actions. This level of control allows developers to tailor the cleaning process to their project's specific needs.

In summary, implementing a cleaning feature in the Bun compiler is a practical and forward-thinking addition that enhances developer productivity, codebase cleanliness, and overall project efficiency. It embodies the principle of "clean code" not only in the source code but also in the output directory, contributing to a more enjoyable and streamlined development experience.

**Comment by silvenon** [comment]
Deleting a directory is simple, no?

```js
import fs from 'node:fs/promises'

await fs.rm('dist', { recursive: true, force: true })
```

It would be different if Bun had a watch API, like esbuild does, but with `bun --watch` or `bun --hot` you can clean with Node.js. How can files end up piling up that way?

Instead, what I'd really like is a _guide_ for creating a simple hot reload server to demonstrate how best to utilize Bun's hot reload feature, server and bundling altogether.

**Reply by guest271314** [reply]
> How can files end up piling up that way?                     

https://github.com/oven-sh/bun/discussions/8078
